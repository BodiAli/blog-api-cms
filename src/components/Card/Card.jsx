import noImage from "../../assets/images/no-image.svg";
import styles from "./Card.module.css";

export default function Card({ post }) {
  console.log(post);

  return (
    <div className={styles.card}>
      <div className={styles.text}>
        <div className={styles.textContainer1}>
          <p className={styles.title}>{post.title}</p>
          <p className={styles.createdAt}>{post.createdAt}</p>
        </div>
        <div className={styles.textContainer2}>
          {post.Topics.length !== 0 &&
            post.Topics.map((topic, i) => {
              return <p key={i}>{topic.name}</p>;
            })}
        </div>
        <div className={styles.textContainer3}>
          <p>{post.likes}</p>
          <p>{post.published ? "published" : "unpublished"}</p>
        </div>
      </div>
      <div className={styles.imageContainer}>
        <img src={post.imgUrl ? post.imgUrl : noImage} alt="Post cover" />
      </div>

      <p className={styles.viewPost}>VIEW</p>
    </div>
  );
}
