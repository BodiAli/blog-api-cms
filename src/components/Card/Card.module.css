.card {
  display: grid;
  grid-template-columns: 2fr 1fr;
  padding: 3rem 0.5rem 0.5rem 0.5rem;
  border-radius: 5px;
  position: relative;
  gap: 10px;
  background-color: rgba(240, 240, 240, 0.211);
}

.text {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

.textContainer1 {
  display: flex;
  gap: 2rem;
  justify-content: space-between;
}

.textContainer1 > .title {
  font-size: 1.5rem;
}

.textContainer2 {
  display: flex;
  align-items: start;
  flex-wrap: wrap;
  gap: 1rem;
  padding: 0 0.5rem;
}

.textContainer2 > p {
  font-weight: 500;
  color: rgba(115, 113, 113, 0.761);
  background-color: rgba(128, 128, 128, 0.246);
  border-radius: 5px;
  padding: 0.3rem 0.6rem;
}

.textContainer3 {
  display: flex;
  justify-content: space-between;
}

.textContainer3 strong {
  font-weight: 500;
}

.textContainer3 > p {
  align-self: center;
}

.textContainer3 > div {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.textContainer4 {
  display: flex;
  gap: 1rem;
}

.textContainer4 > * {
  flex: 1;
  border: none;
  border-radius: 5px;
  color: white;
  padding: 0.3rem 0.6rem;
  font-size: 1rem;
  cursor: pointer;
  transition: background-color 200ms;
}

.textContainer4 > a {
  display: flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  background-color: rgb(72, 130, 180);
  &:hover {
    background-color: rgb(102, 148, 188);
  }
}

.textContainer4 > button:last-child {
  background-color: crimson;
  &:hover {
    background-color: rgb(177, 19, 50);
  }
}

.imageContainer > img {
  width: 100%;
  height: 200px;
  object-fit: cover;
  border-radius: 5px;
}

.viewPost {
  position: absolute;
  top: 5px;
  right: 15px;
  width: 35px;
}

.switch {
  position: relative;
  display: inline-block;
  width: 60px;
  height: 34px;
}

.switch button {
  opacity: 0;
  width: 0;
  height: 0;
}

.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #ccc;
  transition: 400ms;
}

.slider:before {
  position: absolute;
  content: "";
  height: 26px;
  width: 26px;
  left: 4px;
  bottom: 4px;
  background-color: white;
  transition: 400ms;
}

button.published + .slider {
  background-color: #2196f3;
}

button.published + .slider:before {
  transform: translateX(26px);
}

.slider.round {
  border-radius: 34px;
}

.slider.round:before {
  border-radius: 50%;
}

@media (max-width: 550px) {
  .card {
    grid-template-columns: 1fr;
  }

  .viewPost {
    width: 30px;
  }
}
