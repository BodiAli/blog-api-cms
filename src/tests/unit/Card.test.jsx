import { screen, render } from "@testing-library/react";
import { createRoutesStub } from "react-router";
import userEvent from "@testing-library/user-event";
import { toast } from "react-toastify";
import Card from "../../components/Card/Card";

vi.mock("date-fns", () => {
  return {
    formatDistanceToNow: vi.fn(() => "about 1 hour ago"),
  };
});

vi.mock("react-toastify", () => {
  return {
    toast: {
      success: vi.fn(),
      error: vi.fn(),
    },
  };
});

const post = {
  id: 1,
  userId: 1,
  title: "Post 1",
  content: "Post content",
  published: true,
  likes: 5,
  createdAt: "2025-04-27T03:03:02.292Z",
  updatedAt: "2025-04-27T03:03:02.292Z",
  imgUrl: null,
  Topics: [
    { id: 1, name: "Topic 1" },
    { id: 2, name: "Topic 2" },
  ],
};

const requestPostDelete = vi.fn();
const onUpdate = vi.fn();

const updatedPost = {
  ...post,
  published: false,
};

window.fetch = vi.fn(() => {
  return Promise.resolve({
    ok: true,
    status: 200,
    json: vi.fn(() => {
      return Promise.resolve({
        post: updatedPost,
        msg: "Post updated successfully!",
      });
    }),
  });
});

const Stub = createRoutesStub([
  {
    path: "/",
    Component: () => <Card post={post} requestPostDelete={requestPostDelete} onUpdate={onUpdate} />,
  },
]);

describe("Card component", () => {
  test("Should render post data", () => {
    const { container } = render(<Stub />);

    expect(container).toMatchSnapshot();
  });

  test("Should call requestPostDelete with post object", async () => {
    const user = userEvent.setup();
    render(<Stub />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });

    await user.click(deleteButton);

    expect(requestPostDelete).toHaveBeenCalledWith(post);
  });

  test("Should call onUpdate with updated post and show a success toast", async () => {
    const user = userEvent.setup();
    render(<Stub />);

    const publishButton = screen.getByTestId("publish-button");

    await user.click(publishButton);

    expect(onUpdate).toHaveBeenCalledWith(updatedPost);
    expect(toast.success).toHaveBeenCalledWith("Post updated successfully!");
  });

  test("Should show an error toast if fetch fails", async () => {
    window.fetch = vi.fn(() => {
      return {
        ok: false,
        status: 401,
      };
    });

    const user = userEvent.setup();
    render(<Stub />);

    const publishButton = screen.getByTestId("publish-button");

    await user.click(publishButton);

    expect(toast.error).toHaveBeenCalledWith("Failed to update post");
  });
});
