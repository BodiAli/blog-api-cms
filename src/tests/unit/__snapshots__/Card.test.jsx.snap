// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > Should render post data 1`] = `
<div>
  <div
    class="_card_1f2933"
  >
    <div
      class="_text_1f2933"
    >
      <div
        class="_textContainer1_1f2933"
      >
        <p
          class="_title_1f2933"
        >
          Post 1
        </p>
        <p
          class="_createdAt_1f2933"
        >
          about 1 hour ago
        </p>
      </div>
      <div
        class="_textContainer2_1f2933"
      >
        <p>
          Topic 1
        </p>
        <p>
          Topic 2
        </p>
      </div>
      <div
        class="_textContainer3_1f2933"
      >
        <p>
          <strong>
            Likes: 
          </strong>
          5
        </p>
        <div>
          <p>
            Publish:
          </p>
          <label
            class="_switch_1f2933"
          >
            <button
              class="_published_1f2933"
              data-testid="publish-button"
            />
            <span
              class="_slider_1f2933 _round_1f2933"
            />
          </label>
        </div>
      </div>
      <div
        class="_textContainer4_1f2933"
      >
        <a
          data-discover="true"
          href="/users/posts/1/edit"
        >
          Edit
        </a>
        <button>
          Delete
        </button>
      </div>
    </div>
    <div
      class="_imageContainer_1f2933"
    >
      <img
        alt="Post cover"
        src="/src/assets/images/no-image.svg"
      />
    </div>
    <a
      class="_viewPost_1f2933"
      data-discover="true"
      data-testid="view-post"
      href="/users/posts/1"
    >
      <img
        alt="View post"
        src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M12,9A3,3%200%200,1%2015,12A3,3%200%200,1%2012,15A3,3%200%200,1%209,12A3,3%200%200,1%2012,9M12,4.5C17,4.5%2021.27,7.61%2023,12C21.27,16.39%2017,19.5%2012,19.5C7,19.5%202.73,16.39%201,12C2.73,7.61%207,4.5%2012,4.5M3.18,12C4.83,15.36%208.24,17.5%2012,17.5C15.76,17.5%2019.17,15.36%2020.82,12C19.17,8.64%2015.76,6.5%2012,6.5C8.24,6.5%204.83,8.64%203.18,12Z'%20/%3e%3c/svg%3e"
      />
    </a>
  </div>
</div>
`;
